### ISSUE TITLE
When the anchor point is out  of bounds, the whole marker widget disappears

### ISSUE THREAD

**Issue by aytunch** [johnpryan/flutter_map#230]
![simulator screen shot - iphone xr - 2019-02-18 at 23 45 19](https://user-images.githubusercontent.com/6442915/52975740-b211b180-33d7-11e9-91c9-df4e711547db.png)
![simulator screen shot - iphone xr - 2019-02-18 at 23 45 28](https://user-images.githubusercontent.com/6442915/52975743-b63dcf00-33d7-11e9-9e2e-35fc467cbf52.png)

You can see from the second image that left marker is lost with the slightest pan. The moment the little triangle of marker leaves out of any side of the screen, the whole widget disappears. Is there a way to keep it on the map like a safe area or setting a margin for marker?

**Comment by aytunch** [comment]
I had a chance to experience this issue in a real device using hand gestures instead of the mouse and it really bothers me alot to see a bunch of disappearing markers on the screen. Very buggy feeling. I hope there is already a solution or a quick fix. The reason I switched from flutter google maps to flutter maps and mapbox was mainly the power of using custom widget markers(many thanks). And i assume most people will use markers that has a good amount of width like the ones i have instead of default location markers.

**Comment by rayrrr** [comment]
A much easier approach would be to know the width of the marker widget, divide it in half, and then use that number (marker widget * 1/2) as a "padding width" around all 4 edges of the visible area. Then it will behave exactly as we all want, without some complicated mastery of knowing which marker widget is where at all times.

**Comment by johnpryan** [comment]
I think this is an easy fix, thanks for the feedback.

**Comment by aytunch** [comment]
@johnpryan John, were you able to look into this problem? It would be great to have that working:)

**Comment by aytunch** [comment]
> I think this is an easy fix, thanks for the feedback.

@johnpryan Sorry for reminding. I am sure you are very busy. I am reminding again since you said this requires an easy fix. I want my big markers not to disappear:D

**Comment by aytunch** [comment]
@johnpryan Do you think wrapping `FlutterMap` inside of an `SizedBox` which is larger than the screen size or just wrapping it in a container with margin/padding would solve this issue? And also would it have any side effects in terms of for i.e. performance?

**Comment by johnpryan** [comment]
I have not found time to think about the approach. If you have any ideas I would be happy to review a pull request.

**Comment by aytunch** [comment]
Unfortunately all the ideas I have involve higher level solutions instead of deeper level ones. So more like workarounds. Which build function is responsible for drawing markers? Can you send a link for the exact line? Also I wonder how a similar situation with `CircleLayer`, `PolylineLayer`, `PolygonLayer`, `OverlayImageLayer` works just out of the box? Or do they? I never used those however I assume if we have a `PolygonLayer`, I assume it doesn't disappear totally of the center of the polygon is out of the screen bounds. Right?

**Comment by aytunch** [comment]
I found the place: [marker_layer.dart](lib/src/layer/marker_layer.dart)

```
class MarkerLayer extends StatelessWidget {
  final MarkerLayerOptions markerOpts;
  final MapState map;
  final Stream<Null> stream;

  MarkerLayer(this.markerOpts, this.map, this.stream);

  @override
  Widget build(BuildContext context) {
    return StreamBuilder<int>(
      stream: stream, // a Stream<int> or null
      builder: (BuildContext context, AsyncSnapshot<int> snapshot) {
        var markers = <Widget>[];
        for (var markerOpt in markerOpts.markers) {
          var pos = map.project(markerOpt.point);
          pos = pos.multiplyBy(map.getZoomScale(map.zoom, map.zoom)) -
              map.getPixelOrigin();

          var pixelPosX =
              (pos.x - (markerOpt.width - markerOpt.anchor.left)).toDouble();
          var pixelPosY =
              (pos.y - (markerOpt.height - markerOpt.anchor.top)).toDouble();
          // --==== I think in these three lines, we are omitting the offscreen markers ====--
          if (!map.bounds.contains(markerOpt.point)) {
            continue;
          }
          // --====                                                                     ====--   
          markers.add(
            Positioned(
              width: markerOpt.width,
              height: markerOpt.height,
              left: pixelPosX,
              top: pixelPosY,
              child: markerOpt.builder(context),
            ),
          );
        }
        return Container(
          child: Stack(
            children: markers,
          ),
        );
      },
    );
  }
}
```

Instead of using map.bounds in `if (!map.bounds.contains(markerOpt.point)) {...}`, we need to use a new bounds which is extended by half of the markers width in x direction and half of the markers height in y direction. I also know that by the power of you library we can have many different types of marker widgets with different width and height. So it would only make sense to extend map.bounds by user defined `widthextend` and `heightextend`. 

If you can tell me how to extend a `mapBounds` with screen pixel values, I will be happy to try it out and make a PR for you to review.

**Comment by aytunch** [comment]
Also in addition to new mapBounds, I believe that we might need to update the 
```
return Container(
          child: Stack(
            children: markers,
          ),
        );
```
with Stacks property `overflow: Overflow.visible`

**Comment by johnpryan** [comment]
fixed by https://github.com/johnpryan/flutter_map/pull/313

**Comment by aytunch** [comment]
@johnpryan @lpongetti I am using flutter_map 0.5.4 with Flutter 1.5 stable. And this fix #313 does not seem to accomplish its duty. `containsPartialBounds` and  `_boundsContainsMarker` must have something omitted. I hope you guys can give it another look and fix this. And @johnpryan can you open this issue again?

**Comment by aytunch** [comment]
I upgraded to 0.5.5+2 and it works now:)thanks guys. In [pub.dev](https://pub.dev/packages/flutter_map#-changelog-tab-) it says under 0.5.4 that marker disappearing is resolved which was misleading. we need to change that to 0.5.5+2
